File: rust/day_11/part_1.rs
use std::collections::HashMap;
use std::fs;

fn solve(input: &str) -> i64 {
    let lines: Vec<&str> = input
        .trim()
        .split('\n')
        .map(|line| line.trim())
        .filter(|line| !line.is_empty())
        .collect();

    let mut graph: HashMap<String, Vec<String>> = HashMap::new();
    for line in lines {
        let parts: Vec<&str> = line.split(':').collect();
        let device = parts[0].trim().to_string();
        let outputs: Vec<String> = parts[1]
            .trim()
            .split_whitespace()
            .map(|s| s.to_string())
            .collect();
        graph.insert(device, outputs);
    }

    let mut memo: HashMap<String, i64> = HashMap::new();

    fn count_paths(node: &str, graph: &HashMap<String, Vec<String>>, memo: &mut HashMap<String, i64>) -> i64 {
        if node == "out" {
            return 1;
        }

        if let Some(&count) = memo.get(node) {
            return count;
        }

        if !graph.contains_key(node) {
            return 0;
        }

        let mut total = 0;
        if let Some(neighbors) = graph.get(node) {
            for neighbor in neighbors {
                total += count_paths(neighbor, graph, memo);
            }
        }

        memo.insert(node.to_string(), total);
        total
    }

    count_paths("you", &graph, &mut memo)
}

fn main() {
    let input = fs::read_to_string("../inputs/day_11.txt").unwrap();
    let result = solve(&input);
    println!("{}", result);
}

File: go/day_11/part_1.go
package main

import (
	"fmt"
	"os"
	"strings"
)

func solve(input string) int {
	lines := []string{}
	for _, line := range strings.Split(strings.TrimSpace(input), "\n") {
		line = strings.TrimSpace(line)
		if line != "" {
			lines = append(lines, line)
		}
	}

	graph := map[string][]string{}
	for _, line := range lines {
		parts := strings.Split(line, ":")
		device := strings.TrimSpace(parts[0])
		outputs := strings.Fields(strings.TrimSpace(parts[1]))
		graph[device] = outputs
	}

	memo := map[string]int{}

	var count_paths func(node string) int
	count_paths = func(node string) int {
		if node == "out" {
			return 1
		}

		if val, ok := memo[node]; ok {
			return val
		}

		neighbors, ok := graph[node]
		if !ok {
			return 0
		}

		total := 0
		for _, neighbor := range neighbors {
			total += count_paths(neighbor)
		}

		memo[node] = total
		return total
	}

	return count_paths("you")
}

func main() {
	input, err := os.ReadFile("../inputs/day_11.txt")
	if err != nil {
		panic(err)
	}
	result := solve(string(input))
	fmt.Println(result)
}
